numpy/core/include/numpy/npy_common.h

```diff
@@ -359,12 +359,11 @@ typedef unsigned char npy_bool;
 
 
 #if NPY_SIZEOF_LONGDOUBLE == NPY_SIZEOF_DOUBLE
-        typedef double npy_longdouble;
         #define NPY_LONGDOUBLE_FMT "g"
 #else
-        typedef long double npy_longdouble;
         #define NPY_LONGDOUBLE_FMT "Lg"
 #endif
+typedef long double npy_longdouble;
```

Comment:
Can I revisit this one?   @Carlkl pointed out that this is a rather significant change in behavior.  Previously, for platforms and compilers where `double == long double`, at *Numpy compile time*, such as MSVC, this would give `npy_longdouble` as `double`.  This is useful in the case when we are compiling *using* Numpy, but with a compiler where it is *not* true that `double == long double`.  One example is compiling Scipy with mingw-w64, where, by default `long double` is Float80.   Could y'all say more about why this change is needed?  Why do we need to force `npy_longdouble` to be the *compiler* long double, rather than *Numpy* `long double`?

Reply:
Sorry, I think this was my mistake in not looking carefully enough at the PR. I agree we should back out this change.

Reply:
this change was needed to pass the c++ build when `double == long double`, since undefined cast causes compiling errors on C++, also remove the massive cast warnings with C sources.

Reply:
this pr also linked with #20210

Reply:
I completely believe that this change fixed the C++ build, but this is a backwards-incompatible change in the Numpy include headers used by all projects compiling against Numpy, so I do think we need to devote some serious thought to whether this is the right way to fix the C++ build.

Reply:
> Previously, for platforms and compilers where `double == long double`, at _Numpy compile time_, such as MSVC, this would give `npy_longdouble` as `double`. This is useful in the case when we are compiling _using_ Numpy, but with a compiler where it is _not_ true that `double == long double`.

Can we spell the broken workflow(s) out a bit more (sorry, this stuff is confusing)? Is the broken case "build NumPy with MSVC, then SciPy with Mingw"? And does "build both NumPy and SciPy with Mingw" still work?

Reply:
What we need is an additional branch like that:

```C
#if defined __MINGW32__ && __LDBL_DIG__ != __DBL_DIG__
  typedef double npy_longdouble;
#else
  typedef long double npy_longdouble;
#endif
```

explanation: works in case of using the mingw-w64 toolchain regardless if gcc is used with an additional `-mlong-double-64` switch or not.  This would also correctly work for the the Msys2 toolchains, where `long double` is defined as extended precision (FLOAT80).

Reply:
Given that it fixed the `universal2` build, we can't just back out this change. It'd be great to figure out what the right thing to do is though here, because it would be great to not release 1.22.0 with something that's going to prevent building SciPy with Mingw.